core/src/main/java/org/apache/calcite/sql2rel/TopDownGeneralDecorrelator.java

```diff
@@ -0,0 +1,1009 @@
+      List<RexNode> leftJoinConditions = new ArrayList<>();
+      int freeVarsIndex = 0;
+      for (CorDef corDef : corDefs) {
+        RexNode notDistinctFrom =
+            builder.isNotDistinctFrom(
+                builder.field(2, 0, freeVarsIndex),
+                builder.field(2, 1, requireNonNull(corDefOutputs.get(corDef))));
+        leftJoinConditions.add(notDistinctFrom);
+
+        corDefOutputs.put(corDef, freeVarsIndex++);
+      }
+      builder.join(JoinRelType.LEFT, leftJoinConditions);
+
+      // rewrite COUNT to case when
+      List<RexNode> aggCallProjects = new ArrayList<>();
+      final int aggCallStartIndex =
+          dedupFreeVarsNode.r.getRowType().getFieldCount() + newGroupSet.cardinality();
+      for (int i = 0; i < permutedAggCalls.size(); i++) {
+        int index = aggCallStartIndex + i;
+        SqlAggFunction aggregation = permutedAggCalls.get(i).getAggregation();
+        if (aggregation instanceof SqlCountAggFunction) {
+          RexNode caseWhenRewrite =
+              builder.call(
+                  SqlStdOperatorTable.CASE,
+                  builder.isNotNull(builder.field(index)),
+                  builder.field(index),
+                  builder.literal(0));
+          aggCallProjects.add(caseWhenRewrite);
+        } else {
+          aggCallProjects.add(builder.field(index));
+        }
+      }
+      List<RexNode> projects =
+          new ArrayList<>(builder.fields(ImmutableBitSet.range(0, aggCallStartIndex)));
+      projects.addAll(aggCallProjects);
+      newAggregate = builder.project(projects).build();
+
+
+      for (Map.Entry<Integer, Integer> entry : oldToNewOutputs.entrySet()) {
+        int value = requireNonNull(entry.getValue());
+        entry.setValue(value + corDefs.size());
+      }
+    }
+    UnnestQuery unnestQuery
+        = new UnnestQuery(aggregate, newAggregate, corDefOutputs, oldToNewOutputs);
+    mapRelToUnnestQuery.put(aggregate, unnestQuery);
+    return newAggregate;
+  }
+
+  public RelNode unnestInternal(Sort sort) {
+    RelNode newInput = unnest(sort.getInput());
+    UnnestQuery inputInfo =
+        requireNonNull(mapRelToUnnestQuery.get(sort.getInput()));
+    Mappings.TargetMapping targetMapping =
+        Mappings.target(
+            inputInfo.oldToNewOutputs,
+            inputInfo.oldRel.getRowType().getFieldCount(),
+            inputInfo.r.getRowType().getFieldCount());
+    RelCollation shiftCollation = sort.getCollation().apply(targetMapping);
+    builder.push(newInput);
+
+    if (!sort.collation.getFieldCollations().isEmpty()
+        && (sort.offset != null || sort.fetch != null)) {
+      // the Sort with ORDER BY and LIMIT or OFFSET have to be changed during rewriting because
+      // now the limit has to be enforced per value of the outer bindings instead of globally.
+      // It can be rewritten using ROW_NUMBER() window function and filtering on it,
+      // see section 4.4 in paper
```

Comment:
You need to cite both papers, the original one does not contain the sort, this algorithm is only in the second paper